### PR TITLE
Cleanup gumbel_softmax

### LINKED ISSUE #12643: [feature request] Gumbel Softmax for 3D tensor
## 🚀 Feature

I want F.gumbel_softmax to support not only 2D tensors but also 3D or 4D tensors.

## Motivation

I want to apply Gumbel Softmax to 3D tensors.

## Additional context

I think it is easy to make Gumbel Softmax available for Tensors of any number of dimensions.
Are there any reasons why you haven't done yet?

cf.
https://github.com/pytorch/pytorch/blob/a98958d3bdafd4d20e53479b7b070127628a2249/torch/nn/functional.py#L1043

### PR DESCRIPTION
**Pull request by ragulpr**
Fixes #12643, amends to #3341.

- Allow multidimensional input ~~(but apply softmax over `dim=-1`)~~ with `dim` argument
- Cleaner: Less lines of code
- Faster (1.32x speedup vs original, 2x speedup vs using `torch.Distributions`)
- Small fixes in docstring
- Remove some references in docstring. Was the linked (excellent) ipynb the first to do the straight-through trick? Instead, I propose changing to reference to the two papers most known for it.
- Add deprecationwarning for `eps`. It's not needed anymore.
- Initial commit keeps some code alternatives commented to exploit CI

## Controversies:
- As of discussion when `gumbel_softmax` was added (#3341), this was merged into `torch.nn.functional` before all the work with `Distributions` and `Pyro`, and there will probably be multiple other best practices for this in the future. 
I've tested building using the `Distributions`-api, but it was too slow, see below.

I therefore propose not using `Distributions` to keep it fast and simple, but adding a comment in docstring that `gumbel_softmax` may be deprecated in the future. 

### Build on `torch.distributions.RelaxedOneHotCategorical`?
```
dist = torch.distributions.RelaxedOneHotCategorical(temperature=tau, logits=logits, validate_args=False)
y_soft = dist.rsample()
```

Pros:
* Built using tricks like `logsumexp` etc
* Explicitly uses `torch.distributions.utils._finfo` to avoid overflow (old implementation had an `eps` flag)
* Maintained for this exact purpose.

Cons:
* Very slow. Construction of distribution adds overhead see timings below. May be solved in future with speedups of `TransformedDistribution` and `Distribution`.
* Assumes which `dim` to apply softmax over.

### Build on `torch.exponential_()` (as proposed)?
```
    y_soft = logits.new(logits.shape)
    y_soft = (logits - y_soft.exponential_().log()) / tau  # Gumbel noise
    y_soft = y_soft.softmax(dim)  # Gumbel softmax noise
```
Pros:
* Faster

## Timings:
```
    import time
    start = time.time()
    num_draws = 1000000
    logits = torch.randn(1,3)

    for draw in range(num_draws):
        y_draw = gumbel_softmax(logits, hard=True)
        counts = counts + y_draw
    print(end - start)        

## torch.nn.functional.gumbel_softmax()
>> 12.995795965194702

## Using exponential_() as this commit
>> 7.658372640609741

## Using RelaxedOneHotCategorical
>> 20.3382670879364
````

## TODO
Decide on which path to chose. I'll commit in changes to the unit tests in a while to show that it passes both old tests and new tests. I'll also remove the commented code about `RelaxedOneHotCategorical`

### CODE REVIEW

**Comment by ragulpr** [comment]
TODO:
- [ ] Flake8
- [ ] Add new tests
- [ ] Decide on whether to use `torch.Distributions`-api or `.exponential_()`

**Comment by ragulpr** [comment]
Only error with original test was caused by flake8 (hindering other pull requests to merge currently)
```
$ flake8
./torch/nn/functional.py:1173:121: E501 line too long (130 > 120 characters)
./torch/nn/functional.py:1247:121: E501 line too long (130 > 120 characters)
./torch/nn/functional.py:1248:121: E501 line too long (123 > 120 characters)
```

**Comment by ragulpr** [comment]
I think this is ready to be merged. 

* It worked with the original tests, 
* the current tests are more rigorous.
* the api has not changed so no breaking changes. 

Here is a comparison between implementations with full training example https://gist.github.com/ragulpr/1f88aee9fcc585600280ba2d6f323368

**Comment by ragulpr** [comment]
Anyone who could review? I know @soumith reviewed original PR 👍

**Comment by alicanb** [comment]
It would be nice if we keep implementations of [`RelaxedOneHotCategorical.rsample`](https://github.com/pytorch/pytorch/blob/99ce499bfecd4ed0e87cf60d87ee483c6383d95c/torch/distributions/relaxed_categorical.py#L68) and `F.gumbel_softmax` consistent.

**Comment by ragulpr** [comment]
Thanks @alicanb I've been looking into how that would be possible. I think it's two things; consistent in values or consistent in style. Long story short, the line-by-line translation using [reference implementation](https://github.com/pytorch/pytorch/blob/99ce499bfecd4ed0e87cf60d87ee483c6383d95c/torch/distributions/relaxed_categorical.py#L68)

#### Alt a)
```
    uniforms = clamp_probs(torch.rand(logits.shape, dtype=logits.dtype, device=logits.device))
    gumbels = -((-(uniforms.log())).log()) # -log(exponential) = gumbel(0,1)
    scores = (logits + gumbels) / tau # gumbel(logits,tau)
    y_soft = scores - scores.logsumexp(dim=dim, keepdim=True)
    # need to apply transform(x) as it's a TransformedDistribution
    # https://github.com/pytorch/pytorch/blob/99ce499bfecd4ed0e87cf60d87ee483c6383d95c/torch/distributions/transformed_distribution.py#L99
    y_soft = y_soft.exp() 

14.4 s ± 86 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
But this is going to be slow and `logsumexp` unnecessary. To speed this up and leave a smaller [memory footprint](https://gist.github.com/ragulpr/f35513dda57b6f312e0eef107f1f4b2d)  by reusing `gumbels` variable we get below. This is still consistent in values (generates same thing)

#### Alt b)
```
    # merge logsumexp and exp into softmax, reuse variable names
    uniforms = clamp_probs(torch.rand(logits.shape, dtype=logits.dtype, device=logits.device))
    gumbels = -((-(uniforms.log())).log()) # -log(~Exp(1)) = ~Gumbel(0,1)
    gumbels = (logits + gumbels) / tau # ~Gumbel(logits,tau)
    y_soft = gumbels.softmax(dim)

11.5 s ± 130 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Third alternative is to impose the more legible 'style' of relaxed_categorical rsample implementation but use `exponential_()` and `.new()`. This is slightly different in values as it doesn't `clamp_probs` as aggressively.

#### Alt c)
```
    gumbels = -logits.new(logits.shape).exponential_().log() # ~Gumbel(0,1)
    gumbels = (logits + gumbels) / tau  # ~Gumbel(logits,tau)
    y_soft = gumbels.softmax(dim)  # Gumbel softmax noise

11.1 s ± 58.7 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

For reference, this was what I committed;

#### Alt d)
```
    y_soft = logits.new(logits.shape)
    y_soft = (logits - y_soft.exponential_().log()) / tau  # Gumbel noise
    y_soft = y_soft.softmax(dim)  # Gumbel softmax noise

10.6 s ± 59 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

And using finally using `RelaxedOneHotCategorical` without inlining
```
    dist = torch.distributions.RelaxedOneHotCategorical(
        temperature=tau, logits=logits, validate_args=False)
    y_soft = dist.rsample()

21.2 s ± 326 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

In the end, if we want to mimic in values I propose going with *alt b)* otherwise *alt c)*

File: torch/nn/functional.py [hunk 1]
```diff
@@ -1003,72 +1003,53 @@ def softmax(input, dim=None, _stacklevel=3, dtype=None):
...
+    Samples from the `Gumbel-Softmax distribution`_ and optionally discretizes.
 
     Args:
-      logits: `[batch_size, num_features]` unnormalized log probabilities
+      logits: `[..., num_features]` unnormalized log probabilities
       tau: non-negative scalar temperature
       hard: if ``True``, the returned samples will be discretized as one-hot vectors,
             but will be differentiated as if it is the soft sample in autograd
+      eps: Deprecated parameter to avoid log(0). Now handled elsewhere.
+      dim (int): A dimension along which softmax will be computed.
```
Comment (ezyang) [inline torch/nn/functional.py]: Maybe would be better to elide eps from the docs, and specify that dims is keyword only.

File: torch/nn/functional.py [hunk 2]
```diff
@@ -1003,72 +1003,53 @@ def softmax(input, dim=None, _stacklevel=3, dtype=None):
...
-        # this cool bit of code achieves two things:
-        # - makes the output value exactly one-hot (since we add then
-        #   subtract y_soft value)
-        # - makes the gradient equal to y_soft gradient (since we strip
-        #   all other gradients)
-        y = y_hard - y_soft.detach() + y_soft
+        # Straight through.
+        index = y_soft.max(dim, keepdim=True)[1]
+        y_hard = logits.new_zeros(logits.shape).scatter_(dim, index, 1.0)
+        return y_hard - y_soft.detach() + y_soft
```
Comment (ezyang) [inline torch/nn/functional.py]: Aw, I liked the old comment ;)

File: torch/nn/functional.py (hunk 1 above)
Comment (ezyang) [reply torch/nn/functional.py]: dims should say what the default is; as spelled right now it implies the argument is mandatory (which it is not)

**Comment by ezyang** [comment]
@alicanb Do you have an opinion about the implementation strategies?

File: torch/nn/functional.py (hunk 1 above)
Comment (ragulpr) [reply torch/nn/functional.py]: Fixed with a02c8b9 and ea1b998

**Comment by ragulpr** [comment]
It seems like it's failing due to CI-error, and I guess I'm not authorized to call on pytorchbot

File: test/test_nn.py [hunk 3]
```diff
@@ -1951,65 +1951,62 @@ def test_softmax_dtype(self, dtype=torch.float):
...
-    def _test_gumbel_softmax_st(self, cuda, dtype=torch.float):
-        th = torch.cuda if cuda else torch
-        """
-        Things we might want to check:
-        - if we make various draws, do we get different one-hot values?
-        - is the proportion approximately in line with the softmax values?
-        - with hard, is it one-hot?
-        - with hard, is there still a gradient?
-        """
+    def _test_gumbel_softmax_st_shapes(self, dtype, shape, dim, count_expected):
```
[the remark this answers is not in the document]
Comment (ragulpr) [reply test/test_nn.py]: This was fixed with 894612f6079c74c45356823b56001359cb8652c9

**Comment by ragulpr** [comment]
@ezyang , I commited what I called "alternative b)" and made updates according to your comments! I think it's ready for merge.

**Comment by ezyang** [comment]
I'm not sure what's going on with the failures, because the mentioned code doesn't show up in the PR at all. Pushed a merge to master and trying again.

**Comment by ezyang** [comment]
Unfortunately, changes in JIT mean that this patch no longer works, because `clamp_probs` is not scriptable.

@wanchaol could you take a look at this?

**Comment by wanchaol** [comment]
Yes this will make jit standard lib break , @ragulpr why we ended up choosing Alt b) instead of Alt c) given that the Alt c) looks simpler and faster?

**Comment by ragulpr** [comment]
Sorry for slow activity here, got a little bit taken back by all JIT-errors. 

The reason for choosing b)
```
    uniforms = torch.rand_like(logits)
    uniforms = torch.distributions.utils.clamp_probs(uniforms)
    gumbels = -((-(uniforms.log())).log())  # -log(~Exp(1)) = ~Gumbel(0,1)
    gumbels = (logits + gumbels) / tau  # ~Gumbel(logits,tau)
    y_soft = gumbels.softmax(dim)
```
Instead of c)
```
    gumbels = -logits.new(logits.shape).exponential_().log() # ~Gumbel(0,1)
    gumbels = (logits + gumbels) / tau  # ~Gumbel(logits,tau)
    y_soft = gumbels.softmax(dim)  # Gumbel softmax noise
```

Was because alicanb had the good point that implementation be similar to  `RelaxedOneHotCategorical` and *b)* generates identical noise to `RelaxedOneHotCategorical`. Now that we know it wont work with JIT maybe *b)* is not worth the hassle. I just need to verify first that `.exponential_()` will work with JIT too, in that case that is preferable
